smarts/env/wrappers/standard_obs.py

```diff
@@ -0,0 +1,582 @@
+        plane. Defaults to np.array([0,0,0]) per vehicle. shape=(10,3). 
+        dtype=np.float32.
+    speed:
+        Speed of neighbor vehicles in m/s. Defaults to np.array([0]) per
+        vehicle. shape=(10,). dtype=np.float32.
+    """
+    ogm: Optional[np.ndarray] = None
+    """Occupancy grid map. Map is binary, with 255 if a cell is occupied, else
+    0. dtype=np.uint8."""
+    rgb: Optional[np.ndarray] = None
+    """RGB image, from the top view, with ego vehicle at the center.
+    shape=(height, width, 3). dtype=np.uint8."""
+    ttc: Optional[Dict[str, np.ndarray]] = None
+    """Time and distance to collision. Enabled only if both `waypoints` and
+    `neighborhood_vehicles` attributes are enabled in AgentInterface.
+    
+    angle_error:
+        Angular error in radians [-pi, pi]. shape=(1,). dtype=np.float32.
+    distance_from_center:
+        Distance of vehicle from lane center in meters. shape=(1,). 
+        dtype=np.float32.
+    dtc:
+        Distance to collision on the right lane (`dtc[0]`), current lane 
+        (`dtc[1]`), and left lane (`dtc[2]`). If no lane is available, to the 
+        right or to the left, default value of 0 is padded. shape=(3,). 
+        dtype=np.float32.
+    ttc:
+        Time to collision on the right lane (`ttc[0]`), current lane
+        (`ttc[1]`), and left lane (`ttc[2]`). If no lane is available,
+        to the right or to the left, default value of 0 is padded. shape=(3,).
+        dtype=np.float32.
+    """
+    waypoints: Optional[Dict[str, np.ndarray]] = None
+    """Feature array of 20 waypoints ahead or in the mission route, from the 
+    nearest 4 lanes. If lanes or waypoints ahead are insufficient, default 
+    values are padded.
+    
+    heading:
+        Lane heading angle at a waypoint in radians [-pi, pi]. Defaults to
+        np.array([0]) per waypoint. shape=(4,20). dtype=np.float32.
+    lane_index:
+        Lane number at a waypoint. Defaults to np.array([0]) per waypoint.
+        shape=(4,20). dtype=np.int8.
+    lane_width:
+        Lane width at a waypoint in meters. Defaults to np.array([0]) per
+        waypoint. shape=(4,20). dtype=np.float32.
+    pos:
+        Coordinate of a waypoint. Defaults to np.array([0,0,0]). 
+        shape=(4,20,3). dtype=np.float32.
+    speed_limit:
+        Lane speed limit at a waypoint in m/s. shape=(4,20). dtype=np.float32.
+    """
+
+
+class StandardObs(gym.ObservationWrapper):
+    """Converts SMARTS observations to standardized gym-compliant observations
+    and returns `StdObs`. The observation set returned depends on the features
+    enabled via AgentInterface.
+
+    Note: To use StandardObs wrapper, all agents must have the same
+    AgentInterface attributes.
+    """
+
+    def __init__(self, env: gym.Env):
+        """
+        Args:
+            env (gym.Env): SMARTS environment to be wrapped.
+
+        Raises:
+            AssertionError: If all agents do not have the same AgentInterface.
```

Comment:
It occurs to me also that the agent interface adaptors are also potentially incompatible with this wrapper.

Reply:
`FormatObs` wraps the original smarts observations controlled by the `AgentInterface`. Hence, the base environment fed to `FormatObs` should not use `AgentSpec.observation_adapter`. However, `AgentSpec.observation_adapter` can be used in subsequent environments derived and built upon the output environment of the `FormatObs` wrapper.

Following from https://github.com/huawei-noah/SMARTS/pull/1248#discussion_r791060268 and https://github.com/huawei-noah/SMARTS/pull/1248#discussion_r794074825, `FormatObs` is primarily going to be applied only inside gym registration itself. Hence, we can apply the `AgentSpec.observation_adapter`, if any is provided, after `FormatObs` wrapper inside the gym registration. Moreover, once the gym register returns an environment, the users are free to use `AgentSpec.observation_adapter` to modify the observations in an agent-specific manner.

With the above being said, I think observation modifier should be inside the agent and not inside the environment.